grpc-api/src/main/java/io/deephaven/grpc_api/util/GrpcServiceOverrideBuilder.java

```diff
@@ -175,33 +212,80 @@ public ServerServiceDefinition build() {
+            this.sessionService = sessionService;
+        }
+        public ServerCalls.ServerStreamingMethod<ReqT, RespT> open() {
+            return this::invokeOpen;
+        }
+        public ServerCalls.UnaryMethod<ReqT, NextRespT> next() {
+            return this::invokeNext;
+        }
 
-        public OpenBrowserStreamMethod(final Delegate<ReqT, RespT> delegate) {
-            this.delegate = delegate;
+        public void invokeOpen(ReqT request, StreamObserver<RespT> responseObserver) {
+            GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+                StreamData streamData = StreamData.STREAM_DATA_KEY.get();
+                SessionState session = sessionService.getCurrentSession();
+                if (streamData == null) {
+                    throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "no x-deephaven-stream headers, cannot handle open request");
+                }
+
+                BrowserStream<ReqT> browserStream = factory.create(session, responseObserver);
+                browserStream.onMessageReceived(request, streamData);
+
+                if (!streamData.isHalfClose()) {
+                    // if this isn't a half-close, we should export it for later calls - if it is, the client won't send more messages
+                    session.newExport(streamData.getRpcTicket())
+                            .onError(responseObserver::onError)
+                            .submit(() -> browserStream);
+                }
+            });
         }
 
-        @Override
-        public void invoke(final ReqT request, final StreamObserver<RespT> responseObserver) {
-            final ServerCallStreamObserver<RespT> serverCall = (ServerCallStreamObserver<RespT>) responseObserver;
-            serverCall.disableAutoInboundFlowControl();
-            serverCall.request(Integer.MAX_VALUE);
-            delegate.doInvoke(request, responseObserver);
+        public void invokeNext(ReqT request, StreamObserver<NextRespT> responseObserver) {
+            StreamData streamData = StreamData.STREAM_DATA_KEY.get();
+            GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+                final SessionState session = sessionService.getCurrentSession();
+
+                if (streamData == null) {
+                    throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "no x-deephaven-stream headers, cannot handle open request");
+                }
+
+                final SessionState.ExportObject<BrowserStream<ReqT>> browserStream =
+                        session.getExport(streamData.getRpcTicket());
+
+                session.nonExport()
+                        .require(browserStream)
+                        .onError(responseObserver::onError)
+                        .submit(() -> {
+                            browserStream.get().onMessageReceived(request, streamData);
+                            responseObserver.onNext(null);//TODO simple response payload
```

Comment:
Perhaps your factory could also give you the "default instance" of a NextRespT?

Reply:
Not necessary right? If we deliberately send null, zero bytes are sent, and the receiver marshaller will use the default value?

Need to drop the TODO though